drivers/ruby/lib/net.rb

```diff
@@ -355,13 +398,23 @@ def em_run(*args, &b)
         raise ArgumentError, "No handler specified."
       end
 
+      async_handler = args[:async_handler]
+      if !async_handler.is_a?(AsyncHandler)
+        raise ArgumentError, "No async handler specified."
+      end
+
       # If the user has defined the `on_state` method, we assume they want states.
       if args[:block].respond_to?(:on_state)
         args[:opts] = args[:opts].merge(include_states: true)
       end
 
-      EM_Guard.register(args[:conn])
-      args[:conn].run(@body, args[:opts], args[:block])
+      async_handler.callback = args[:block]
+      async_handler.connection = args[:conn]
+      async_handler.options = args[:opts]
+
+      async_handler.run do
+        async_handler.connection.run(@body, async_handler.options, async_handler.handler)
```

Comment:
There's some serious class jealousy here. Do you think it would be prudent to make the AsyncHandler run function just take a callback, connection, options and body so we could just collapse these 6 lines?